components/butterfly/src/server.rs

```diff
@@ -1073,10 +1134,30 @@ impl Server {
                                    .map(|stored_term| election.term > stored_term)
                                    .unwrap_or(false);
                 if new_term {
-                    debug!("removing old rumor and starting new election");
-                    self.election_store
-                        .remove_rsw(election.key(), election.id());
-                    self.start_election_rsw_mlr_rhw_msr(&election.service_group, election.term);
+                    if Some(self.member_id())
+                       == self.election_store.lock_rsr().get_member_id(election.key())
+                    {
+                        debug!("I am the leader of previous term!");
+                        debug!("removing old rumor and starting new election");
+                        self.election_store
+                            .remove_rsw(election.key(), election.id());
+                        self.start_election_rsw_mlr_rhw_msr(&election.service_group, election.term);
+                    } else if Some(election.member_id.as_str())
+                              == self.election_store.lock_rsr().get_member_id(election.key())
+                    {
+                        debug!("Received New Term election from the leader. Starting my own to \
+                                merge. Term: {}",
+                               election.term);
+                        self.election_store
+                            .remove_rsw(election.key(), election.id());
+                        trace!("Removed old election.");
+                        self.start_election_rsw_mlr_rhw_msr(&election.service_group, election.term);
+                        trace!("Started new election.");
+                    } else {
+                        warn!("Received a New Term Election, but not from the current leader, \
+                               ignoring!");
+                        return;
+                    }
```

Comment:
In my testing, this `insert_election_rsw_mlr_rhw_msr` code is not consistently re-electing the same leader.  I test with 4 members. Often the leader is elected before all 4 members have started up since only two are needed for quorum. So the chances that the lexagraphically highest member will be leader is not guaranteed. If a laxagraphicaly higher member joins after the leader is elected, when a follower is partitioned and starts a se-election, a different member will be elected.

I think what this method needs is actually something more simple that what is here.  In the block where it sees that it was the leader of the last election, it should have a way of setting the suitability to the max i64 value.

I had tried that and the last leader was always re-elected.

Reply:
Yes, I have seen that - however that re-election happens only first time but not again and again. But I believe yes that can be avoided as well. But the *new term* has to be started by the leader of the previous term, if we ignore that - then a new leader can get re-elected as follows -

A - B - C  
term: 0 C is the leader A and B are followers 

B is partitioned A and C form the quorum  and still things are fine. 

B comes out of partition and has received Ping from A and has marked A as alive but "C" is not alive (even though A said it is alive - because of the incarnation logic). 

A starts election with zero suitability and that rumor reaches B. B joins the election has quorum and determines that  there are 2 members alive and there is quorum (Because it has not yet received `Ping` or `Ack` from `C` yet and never sent a `ProbePing` as `C` was never returned as `Confirmed`) and marks the election as `Finished`. 

So while `C` was legitimately alive, but `B` received the election rumor before it heard from `C`, it resulted in new leader. `Suitability` is not considered when the election is `Finished`.

With higher number of nodes like 4-5, even there - there is re-election only once if that happens because that node joined the election `Late`. I have tested it with 1 and 2 simultaneous node failures for 5 nodes about 50-70 times. With `Suitability` logic even that can be avoided.

Reply:
I was trying to avoid a change to `restart_elections_xxx` code since that is a public API. But I believe it's a fine if we do it by sending `Option<i64>` as `Suitabiliy`.